### ISSUE TITLE
Is there any way to get the CloudEvent wrapper when subscribing to a pubsub topic?

### ISSUE THREAD

**Issue by JasonRodman** [dapr/dapr#6965]
In my implementation I was looking into ways to deduplicate events, thinking I could use the CloudEvent's id property like the event publishing documentation says. I could not find any way to get the envelope on the subscriber side since it appears only the event data is sent in the post to the consumer side. I didn't see it in the headers either. Is this something that there is a configuration setting for in the dapr component? I do not see how to get that data in subscribers, only a brief mention of it here at the bottom of this page. 
https://docs.dapr.io/developing-applications/building-blocks/pubsub/pubsub-cloudevents/

**Comment by Joffreyvl** [comment]
In the dotnet SDK you can set EnableRawPayload on the TopicOptions to get the raw data from you PubSub component.

![image](https://github.com/dapr/dapr/assets/17174541/bf67437c-14ca-481b-9900-04d098f67374)

**Comment by JasonRodman** [comment]
This does not work for me. The documentation says this is when you want to create custom payloads, not cloudevents. I simply want the raw cloudevent rather than the unwrapped event from the payload. Dapr delivers the entire cloudevent wrapper to the endpoint and its middleware unwraps it. But when this option is enabled the dapr sidecar fails to deliver the event to my endpoint with "no matching route for event" in the logs. What did work is instead exclude the .UseCloudEvents call in Program.cs which adds the middleware. Now all events are delivered in their raw form and I can deserialize them into a model.

**Comment by JasonRodman** [comment]
That is one way for sure. DAPR delivers the payload in the cloudevent wrapper to your application and it is unwrapped by the SDK middleware, which is forwarded to your handler as just the event without the wrapper. I simply omitted the `UseCloudEvents()` call in Program.cs and then your handler will get the raw payload. I preferred that approach, so I created my own models to mirror the cloudevent since the SDK CloudEvent models omit a few things. My implementation looks like this:

```cs
public record DaprEventWrapper<TEvent>(TEvent Data, 
    string? PubSubName, string? Topic, string? Type, string? Subject,
    string? DataContentType, string? Source, string? SpecVersion, 
    string? TraceId, string? TraceState, string? TraceParent);
```
I did not need to turn on `TopicOptinos.UseRawPayload` because it was not necessary, and it breaks my routing when I do. I am still a bit frustrated I cannot get the the SDK to allow me to override the cloudevent.id property. It just simply does not work. I have an open issue on that feature at the moment.
